File: LPYFramework/LPYFramework/Category/UIView+Extension.m
//
//  UIView+Extension.m
//  QQMProject
//
//  Created by admin on 15/10/19.
//  Copyright © 2015年 admin. All rights reserved.
//

#import "UIView+Extension.h"

@implementation UIView (Extension)
- (void)setWidth_:(CGFloat)width_
{
    CGRect frame = self.frame;
    frame.size.width = width_;
    self.frame = frame;
}

- (CGFloat)width_
{
    return self.frame.size.width;
}

- (void)setHeight_:(CGFloat)height_
{
    CGRect frame = self.frame;
    frame.size.height = height_;
    self.frame = frame;
}

- (CGFloat)height_
{
    return self.frame.size.height;
}

- (void)setX_:(CGFloat)x_
{
    CGRect frame = self.frame;
    frame.origin.x = x_;
    self.frame = frame;
}

- (CGFloat)x_
{
    return self.frame.origin.x;
}

- (void)setY_:(CGFloat)y_
{
    CGRect frame = self.frame;
    frame.origin.y = y_;
    self.frame = frame;
}

- (CGFloat)y_
{
    return self.frame.origin.y;
}

- (void)setCenterX_:(CGFloat)centerX_
{
    CGPoint center = self.center;
    center.x = centerX_;
    self.center = center;
}

-(CGFloat)centerX_
{
    return self.center.x;
}

- (void)setCenterY_:(CGFloat)centerY_
{
    CGPoint center = self.center;
    center.y = centerY_;
    self.center = center;
}

- (CGFloat)centerY_
{
    return self.center.y;
}

- (void)setSize_:(CGSize)size_
{
    CGRect frame = self.frame;
    frame.size = size_;
    self.frame = frame;
}

- (CGSize)size_
{
    return self.frame.size;
}

+ (BOOL)isShowOnKeyWindow:(UIView *)view
{
    // 将当前View的坐标系转换为窗口坐标系
    // 第二个参数如果写nil，表示的也是UIWindow
    //        CGPoint windowP = [subView convertPoint:subView.frame.origin toView:[UIApplication sharedApplication].keyWindow];
    // 将控件在父控件的坐标系转换为在主窗口的坐标系
    
    // 判断两个矩形是否相交
    //        CGRectIntersectsRect(<#CGRect rect1#>, <#CGRect rect2#>)
    
    UIWindow *window = [UIApplication sharedApplication].keyWindow;
    
    CGRect windowRect = window.bounds;
    CGRect newRect = [view.superview convertRect:view.frame toView:window];
    return view.window == window && !view.hidden && view.alpha > 0.01 && CGRectIntersectsRect(windowRect, newRect);
}

+ (instancetype)viewFromXib
{
    return [[[NSBundle mainBundle] loadNibNamed:NSStringFromClass(self) owner:nil options:nil] lastObject];
}

- (UIViewController *)viewController {
    for (UIView *next = [self superview]; next; next = next.superview) {
        UIResponder *nextResponder = [next nextResponder];
        if ([nextResponder isKindOfClass:[UIViewController class]]) {
            return (UIViewController*)nextResponder;
        }
    }
    return nil;
}

@end

File: 第三方框架改造/pull-to-refresh-master/ESPullToRefreshExample/ESPullToRefreshExample/Custom/Meituan/UIView+Extension.swift
//
//  UIView+Extension.swift
//  QQMProjectSwift
//
//  Created by liupuyan on 2017/5/12.
//  Copyright © 2017年 liupuyan. All rights reserved.
//

import Foundation
import UIKit

extension UIView {
    /**
     * 高度
     */
    var height_: CGFloat {
        get {
            return frame.size.height
        }
        set(newValue) {
            var frame: CGRect = self.frame
            frame.size.height = newValue
            self.frame = frame
        }
    }
    /**
     * 宽度
     */
    var width_: CGFloat {
        get {
            return frame.size.width
        }
        set(newValue) {
            var frame: CGRect = self.frame
            frame.size.width = newValue
            self.frame = frame
        }
    }
    var x_: CGFloat {
        get {
            return frame.origin.x
        }
        set(newValue) {
            var frame: CGRect = self.frame
            frame.origin.x = newValue
            self.frame = frame
        }
    }
    var y_: CGFloat {
        get {
            return frame.origin.y
        }
        set(newValue) {
            var frame: CGRect = self.frame
            frame.origin.y = newValue
            self.frame = frame
        }
    }
    var centerX_: CGFloat {
        get {
            return center.x
        }
        set(newValue) {
            var center: CGPoint = self.center
            center.x = newValue
            self.center = center
        }
    }
    var centerY_: CGFloat {
        get {
            return center.y
        }
        set(newValue) {
            var center: CGPoint = self.center
            center.y = newValue
            self.center = center
        }
    }
    var size_: CGSize {
        get {
            return frame.size
        }
        set(newValue) {
            var frame: CGRect = self.frame
            frame.size = newValue
            self.frame = frame
        }
    }
}


// MARK: - layer
extension UIView {
    class func createGradientLayer(frame: CGRect, startColor: UIColor, endColor: UIColor, gradientAngle: CGFloat) -> CAGradientLayer {
        return createGradientLayer(frame: frame, startColor: startColor, endColor: endColor, gradientAngle: gradientAngle, startPoint: CGPoint(x: 1, y: 1), endPoint: CGPoint(x: 0, y: tan(gradientAngle)))
    }
    
    class func createGradientLayer(frame: CGRect, startColor: UIColor, endColor: UIColor, gradientAngle: CGFloat, startPoint: CGPoint, endPoint: CGPoint) -> CAGradientLayer {
        let gradientLayer: CAGradientLayer = CAGradientLayer()
        gradientLayer.frame = frame
        gradientLayer.startPoint = startPoint
        gradientLayer.endPoint = endPoint
        if gradientAngle > 0 {
            let l: NSNumber = NSNumber(value: Float(tan(gradientAngle)))
            gradientLayer.locations = [l]
        }
        let startColor = startColor.cgColor
        let endColor = endColor.cgColor
        gradientLayer.colors = [startColor, endColor]
        return gradientLayer
    }
}
